src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs

```diff
@@ -218,93 +221,202 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
         {
             Debug.Assert(length >= 0);
 
-            fixed (char* pChars = &searchSpace)
-            {
-                char* pCh = pChars;
-                char* pEndCh = pCh + length;
+            nint offset = (nint)0;
+            nint lengthToExamine = length;
 
-                if (Vector.IsHardwareAccelerated && length >= Vector<ushort>.Count * 2)
+            if (Avx2.IsSupported || Sse2.IsSupported)
+            {
+                // Avx2 branch also operates on Sse2 sizes, so check is combined.
+                if (length >= Vector128<byte>.Count * 2)
                 {
-                    // Figure out how many characters to read sequentially until we are vector aligned
-                    // This is equivalent to:
-                    //         unaligned = ((int)pCh % Unsafe.SizeOf<Vector<ushort>>()) / elementsPerByte
-                    //         length = (Vector<ushort>.Count - unaligned) % Vector<ushort>.Count
-                    const int elementsPerByte = sizeof(ushort) / sizeof(byte);
-                    int unaligned = ((int)pCh & (Unsafe.SizeOf<Vector<ushort>>() - 1)) / elementsPerByte;
-                    length = (Vector<ushort>.Count - unaligned) & (Vector<ushort>.Count - 1);
+                    lengthToExamine = UnalignedCountVector128(ref searchSpace);
                 }
-
-            SequentialScan:
-                while (length >= 4)
+            }
+            else if (Vector.IsHardwareAccelerated)
+            {
+                if (length >= Vector<ushort>.Count * 2)
                 {
-                    length -= 4;
+                    lengthToExamine = UnalignedCountVector(ref searchSpace);
+                }
+            }
 
-                    if (pCh[0] == value)
-                        goto Found;
-                    if (pCh[1] == value)
-                        goto Found1;
-                    if (pCh[2] == value)
-                        goto Found2;
-                    if (pCh[3] == value)
-                        goto Found3;
+        SequentialScan:
+            while (lengthToExamine >= 4)
+            {
+                ref char current = ref Add(ref searchSpace, offset);
```

Comment:
Is the ref arithmetic better than just pinning at the entry-point?

I've seen @jkotas leave comments on similar PRs that the cost of pinning is generally cheap, provided you aren't doing it in a hot loop; and it is generally more readable/less error-prone than the ref arithmetic.

Reply:
It depends on the overall complexity of the method. If the method is simple, I am ok with using the ref arithmetic because of the few extra instructions from pinning slow it down significantly.

I think this one is around the treashold where the ref arithmetic is not justified anymore.

Reply:
Would result in double pinning as this method is called with pinned and unpinned data

Reply:
Also is quite a common called method; better to let the GC a free hand to relocate? 

If the data is not pre-pinned, loosing alignment after compaction is likely to dwarfed by the GC itself?